### PR TITLE
High: CIB: no schema violating oversimplified patching with @id-ref

### PR DESCRIPTION
**Pull request by jnpkrn**
Previously, this input to modification of CIB:

> <primitive ...>
>   <operations>
>     <op id="mySmartFuse-monitor-outputpower" name="monitor" interval="2s">
>       <instance_attributes id="mySmartFuse-outputpower-instanceparams">
>         <nvpair id="mySmartFuse-outputpower-req" name="req" value="outputpower"/>
>       </instance_attributes>
>     </op>
>   </operations>
>   <instance_attributes id="mySmartFuse-params">
>     <nvpair id="mySmartFuse-params-ip" name="ip" value="192.0.2.10"/>
>   </instance_attributes>
>   <instance_attributes id-ref="mySmartFuse-outputpower-instanceparams"/>
> </primitive>

would degenerate into this:

> <primitive ...>
>   <operations>
>     <op id="mySmartFuse-monitor-outputpower" name="monitor" interval="2s">
>       <instance_attributes id="mySmartFuse-outputpower-instanceparams">
>         <nvpair id="mySmartFuse-outputpower-req" name="req" value="outputpower"/>
>       </instance_attributes>
>     </op>
>   </operations>
>   <instance_attributes id="mySmartFuse-params" id-ref="mySmartFuse-outputpower-instanceparams>
>     <nvpair id="mySmartFuse-params-ip" name="ip" value="192.0.2.10"/>
>   </instance_attributes>
> </primitive>

apparently resulting in an invalid configuration instance
(cannot mix @id-ref with anything else incl. @id).

Now, we are more careful about @id-ref so this won't happen.
As mentioned in the in-code comment, it's not the best qualitative
arrangement, but there's no time complexity penalty, either.

### CODE REVIEW

**Comment by kgaillot** [comment]
What do you think of:

    CRM_CHECK(target || parent);

That would be fine with existing usage, and it makes sense to me ... if the caller didn't specify a target, we must have a parent to search/create a target under.

If that's the case, we don't need find_xml_node() here at all, b/c find_entity() with object_id==NULL will have the same effect. So the existing block could just be:

    if (target == NULL) {
        target = find_entity(parent, object_name, object_id);
    }

I'm not clear on how id-ref fits into the picture. Maybe if object_id is NULL, we should check for object_idref, and if that is present, we should do the equivalent of find_entity() by id-ref instead of id.

**Comment by jnpkrn** [comment]
Generally good idea, but one day, it will likely beat my sanity,
it's _never_ something in isolation, more like infinite rabbit hole.

Untested, but it's exercised in the test for parallel #1500, so will
get to it.

**Comment by jnpkrn** [comment]
Fixed last commit's message.

**Comment by jnpkrn** [comment]
Actually, `id` is no special, it would be broken with the last commit otherwise.
Fixed.

**Comment by jnpkrn** [comment]
Fixed a bug/thinko in `find_entity_by_id_or_first`, however there's more
to look at, eventually.  When tracing enabled for `test_tools` in `cts-cli`,
`crm_ticket -t ticketA -r --force` will fail on:

```
xml.c: add_xml_object
CRM_CHECK(safe_str_eq(crm_element_name(target), crm_element_name(update)), ...);

(gdb) p target->name
$2 = (const xmlChar *) 0x16fad40 "node_state"
(gdb) p update->name
$3 = (const xmlChar *) 0x16f8400 "tickets"
```

**Comment by jnpkrn** [comment]
Hm, actually that was a matter of yet another oversight.  Fixed as well.

**Comment by jnpkrn** [comment]
Perhaps `sfind_entity_by_attr_or_name/find_entity_by_attr_or_just_name/`
to make it absolutely clear.

File: lib/common/xml.c [hunk 1]
```diff
@@ -2258,23 +2258,47 @@ find_xml_node(xmlNode * root, const char *search_path, gboolean must_find)
...
+    CRM_CHECK(attr_n == NULL || attr_v != NULL, return NULL);
+
+    for (child = __xml_first_child(parent); child != NULL; child = __xml_next(child)) {
+        /* XXX uncertain if the first check is strictly necessary here */
+        if (node_name == NULL || !strcmp((const char *) child->name, node_name)) {
+            if (attr_n == NULL
+                    || crm_str_eq(crm_element_value(child, attr_n), attr_v, TRUE)) {
+                return child;
             }
+            first_child = child;
```
Comment (kgaillot) [inline lib/common/xml.c]: Did you mean to wrap this with if (first_child == NULL)?

Comment (jnpkrn) [reply lib/common/xml.c]: Heck, indeed, good catch.
The subfix (amendment of the forgotten piece exposed directly in the tests)
that itself went too carelessly wrt. intended semantics.

Comment (jnpkrn) [reply lib/common/xml.c]: In fact, it wasn't meant to be present at all (Friday + mess arising from several branches
worked on in parallel?)

Comment (jnpkrn) [reply lib/common/xml.c]: Of course, now it might make sense to put it all into one big `if`.

**Pull request merged**